### PR TITLE
#903 add units out of service

### LINKED ISSUE #903: Introduce units_out_of_service to schedule maintenance outages
To increase the flexibility and utility of SpineOpt, it would be useful to be able to schedule maintenance outages. A simple way to do this is to add a new variable called `units_out_of_service`. 

At any time, t, `units_available(u, t)` is then `number_of_units(u, t)` + `units_invested_available(u, t)` - `units_out_of_service(u,t)`

Then we have a new parmeter `scheduled_outage_duration` which is the amount of time a unit must be out on maintenance over the optimisation horizon. The idea is that this would be used in a long term model, perhaps along with scheduling of long term storage/hydro.

Then we need a constraint similar to `min_up_time` to ensure that the outage is taken as a continuous block

My understanding is that the most efficient way to implement would be to have the following analogs:

`units_out_of_service` = `units_on`
`units_taken_out_of_service` = `units_started_up`
`units_returned_to_service` = `units_shut_down`
`units_out_of_service_transition` = `unit_state_transition`
`units_out_of_service_contiguous` = `min_up_time`

New constraint:

min_scheduled_outage_duration:
sum(t, units_out_of_service(u,y)) >= scheduled_outage_duration(u) * number_of_units(u)

In a second piece, we can introduce a piecewise linear penalty function into the objective function to maximise the capacity margin (minimise LOLE) over the optimisation window

### PR DESCRIPTION
**Pull request by DillonJ**
Introduces a new variable called `units_out_of_service`. This allows a user to schedule maintenance outages. The creation of the varialble is controlled via `outage_variable_type` which uses the same list as `online_variable_type` and takes the same resolution. `outage_variable_type` defaults to a value of `online_variable_type_none` meaning the default behaviour is that this variable is not created and there is no impact on existing models. A new parameter `scheduled_outage_duration` may be specified and when `outage_variable_type` has a value of anything other than `online_variable_type_none`, the constraint `min_scheduled_outage_duration` will be triggered which will enforce that sum(t, units_out_of_service(t)) >= `scheduled_outage_duration`. 

In an upcoming development, this can be used in conjunction with a pentalty on the capacity margin in each timestep, allowing outages to be scheduled in a way that minimises the impact on capacity adequacy.

Fixes #903

## Checklist before merging
- [x] Documentation is up-to-date
- [x] Unit tests have been added/updated accordingly
- [x] Code has been formatted according to SpineOpt's style
- [x] Unit tests pass

### CODE REVIEW

File: src/data_structure/migration.jl [hunk 1]
```diff
@@ -37,6 +37,10 @@ include("versions/rename_benders_master_to_just_benders.jl")
 include("versions/translate_ramp_parameters.jl")
 include("versions/remove_model_tb_ss.jl")
 
+function units_out_of_service_upgrade(db_url, log_level)
+	true
+end
+
```
Comment (manuelma) [inline src/data_structure/migration.jl:43]: Does this rely on users running with upgrade=true for things to work?

File: src/variables/variable_units_out_of_service.jl [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        :units_out_of_service,
+        units_on_indices;
+        lb=Constant(0),
+        bin=units_out_of_service_bin,
+        int=units_out_of_service_int,
+        fix_value=fix_units_out_of_service,
+        initial_value=initial_units_out_of_service,
+        replacement_value=units_out_of_service_replacement_value,
+        #non_anticipativity_time=units_on_non_anticipativity_time,
+        #non_anticipativity_margin=units_on_non_anticipativity_margin,
```
Comment (manuelma) [inline src/variables/variable_units_out_of_service.jl:69]: Did you forget to remove these commented lines?

File: src/constraints/constraint_min_scheduled_outage_duration.jl [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    t0 = _analysis_time(m)
+    m.ext[:spineopt].constraints[:min_scheduled_outage_duration] = Dict(
+        (unit=u, stochastic_path=s_long, t=t_long) => @constraint(
+            m, 
+            + sum(
+                units_out_of_service[u, s, t] * duration(t)
+                for (u, s, t) in units_on_indices(m; unit=u);
+                init=0
+            ) 
+            >= ( + scheduled_outage_duration[(unit=u, stochastic_scenario=s_long, analysis_time=t0, t=t_long)]
```
Comment (manuelma) [inline src/constraints/constraint_min_scheduled_outage_duration.jl:39]: Maybe we need to break the line after the left parenthesis?

File: src/constraints/constraint_min_scheduled_outage_duration.jl [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+Creates all indices required to include units, stochastic paths and a reference temporal for `add_constraint_scheduled_outage_duration!`
+constraint generation.
+"""
+function constraint_min_scheduled_outage_duration_indices(m::Model)
+    unique(
+        (unit=u, stochastic_path=path, t=t)
+        for u in indices(scheduled_outage_duration)
+        for t in current_window(m)
+        for path in active_stochastic_paths(
+            m, [units_on_indices(m; unit=u); ]
```
Comment (manuelma) [inline src/constraints/constraint_min_scheduled_outage_duration.jl:60]: I think `[a; ]` can be written as just `a`?

File: src/data_structure/benders_data.jl [hunk 5]
```diff
@@ -154,6 +154,7 @@ end
 function _set_sp_solution!(m, k)
     for (name, vals) in get(m.ext[:spineopt].sp_values, k, ())
         var = m.ext[:spineopt].variables[name]
+        var isa VariableRef || continue
```
Comment (manuelma) [inline src/data_structure/benders_data.jl:157]: Nice catch, we should have this one already in master.

File: test/constraints/constraint_unit.jl [hunk 6]
```diff
@@ -184,7 +184,7 @@ function test_constraint_units_available()
         ]
         SpineInterface.import_data(url_in; relationships=relationships, object_parameter_values=object_parameter_values)
         m = run_spineopt(url_in; log_level=0, optimize=false)
-        var_units_available = m.ext[:spineopt].variables[:units_available]
+        var_units_available = m.ext[:spineopt].variables[:units_available]
```
Comment (manuelma) [inline test/constraints/constraint_unit.jl:187]: Just adding white space?

File: src/data_structure/migration.jl (hunk 1 above)
Comment (DillonJ) [reply src/data_structure/migration.jl:43]: not if we can get [(... _default=0)] working

File: test/constraints/constraint_unit.jl [hunk 7]
```diff
@@ -235,6 +270,41 @@ function test_constraint_unit_state_transition()
     end
 end
 
+function test_units_out_of_service_transition()
+    @testset "constraint_unit_state_transition" begin
```
Comment (manuelma) [inline test/constraints/constraint_unit.jl:274]: Is it ok to reuse a @testset name? I am not sure they both get executed in that case - could you check?

File: test/constraints/constraint_unit.jl [hunk 8]
```diff
@@ -947,6 +1017,92 @@ function test_constraint_min_up_time()
...
+            
+            scenarios = [
+                [stochastic_scenario(:parent)] 
+                repeat([stochastic_scenario(:child)], 4)
+            ]
+            time_slices = time_slice(m; temporal_block=temporal_block(:hourly))
+
+            vars_u_oos = [var_units_out_of_service[unit(:unit_ab), s, t] for (s, t) in zip(scenarios, time_slices)]
+            expected_con = @build_constraint(sum(vars_u_oos) >= scheduled_outage_duration_minutes/60 )
+            @show expected_con
```
Comment (manuelma) [inline test/constraints/constraint_unit.jl:1097]: Maybe remove the @show?

File: src/variables/variable_units_out_of_service.jl (hunk 2 above)
Comment (DillonJ) [reply src/variables/variable_units_out_of_service.jl]: Fixed this in [latest commit](https://github.com/spine-tools/SpineOpt.jl/commit/21097e41186dd36e83125e591bcfd34989910eeb)

File: src/constraints/constraint_min_scheduled_outage_duration.jl (hunk 3 above)
Comment (DillonJ) [reply src/constraints/constraint_min_scheduled_outage_duration.jl]: Fixed this in [latest commit](https://github.com/spine-tools/SpineOpt.jl/commit/21097e41186dd36e83125e591bcfd34989910eeb)

File: src/constraints/constraint_min_scheduled_outage_duration.jl (hunk 4 above)
Comment (DillonJ) [reply src/constraints/constraint_min_scheduled_outage_duration.jl]: Ah yes! Done in [latest commit](https://github.com/spine-tools/SpineOpt.jl/commit/21097e41186dd36e83125e591bcfd34989910eeb)

File: test/constraints/constraint_unit.jl (hunk 6 above)
Comment (DillonJ) [reply test/constraints/constraint_unit.jl]: Fixed in [latest commit](https://github.com/spine-tools/SpineOpt.jl/commit/21097e41186dd36e83125e591bcfd34989910eeb)

File: test/constraints/constraint_unit.jl [hunk 9]
```diff
@@ -201,6 +201,41 @@ function test_constraint_units_available()
             @test _is_constraint_equal(observed_con, expected_con)
         end
     end
+    @testset "constraint_units_available_units_unavailable" begin
```
[the remark this answers is not in the document]
Comment (DillonJ) [reply test/constraints/constraint_unit.jl:206]: Done in [latest commit](https://github.com/spine-tools/SpineOpt.jl/commit/21097e41186dd36e83125e591bcfd34989910eeb)

File: test/constraints/constraint_unit.jl (hunk 7 above)
Comment (DillonJ) [reply test/constraints/constraint_unit.jl]: Fix this just to be sure in [latest commit](https://github.com/spine-tools/SpineOpt.jl/commit/21097e41186dd36e83125e591bcfd34989910eeb)

File: test/constraints/constraint_unit.jl (hunk 8 above)
Comment (DillonJ) [reply test/constraints/constraint_unit.jl]: Thanks, done in [latest commit](https://github.com/spine-tools/SpineOpt.jl/commit/21097e41186dd36e83125e591bcfd34989910eeb)

**Comment by manuelma** [comment]
It is a bit unfortunate that we couldn't fix the issue with the dummy migration script here... but great job anyways!

**Comment by DillonJ** [comment]
> It is a bit unfortunate that we couldn't fix the issue with the dummy migration script here... but great job anyways!
> 
Ah - I actually created the dummy script and that's how I updated the examples... but perhaps I didn't commit the file

**Comment by manuelma** [comment]
> > It is a bit unfortunate that we couldn't fix the issue with the dummy migration script here... but great job anyways!
> 
> Ah - I actually created the dummy script and that's how I updated the examples... but perhaps I didn't commit the file @manuelma should I just create the dummy script and push?

Not sure @DillonJ - it seems to work after restarting julia... it is just a bit unfortunate that we introduced a new version for something we could have fixed better.

**Comment by manuelma** [comment]
Getting this now, maybe an issue with the replacement value?
```
- [constraint_units_available]
ERROR: LoadError: MethodError: no method matching zero(::Nothing)

Closest candidates are:
  zero(::Union{Type{P}, P}) where P<:Dates.Period
   @ Dates C:\Julia\Julia-1.9.3\share\julia\stdlib\v1.9\Dates\src\periods.jl:51
  zero(::FillArrays.AbstractZeros{T, N}) where {T, N}
   @ FillArrays C:\Users\Manuel_Marin\.julia\packages\FillArrays\oXkMk\src\FillArrays.jl:599
  zero(::JuMP.GenericQuadExpr)
   @ JuMP C:\Users\Manuel_Marin\.julia\packages\JuMP\R53zo\src\quad_expr.jl:126
  ...

Stacktrace:
  [1] iszero(x::Nothing)
    @ Base .\number.jl:42
  [2] _diff_call(x::Nothing, y::SpineInterface.Call)
    @ SpineInterface C:\Users\Manuel_Marin\Codes\SpineInterface.jl\src\base.jl:248
  [3] -(x::SpineInterface.Call, y::SpineInterface.Call)
    @ SpineInterface C:\Users\Manuel_Marin\Codes\SpineInterface.jl\src\base.jl:242
  [4] _sum_call(args::Vector{Union{Nothing, SpineInterface.Call}})
    @ SpineInterface C:\Users\Manuel_Marin\Codes\SpineInterface.jl\src\base.jl:203
  [5] +(x::SpineInterface.Call, y::SpineInterface.Call)
    @ SpineInterface C:\Users\Manuel_Marin\Codes\SpineInterface.jl\src\base.jl:175
  [6] +(lhs::SpineInterface.Call, rhs::JuMP.GenericAffExpr{SpineInterface.Call, JuMP.VariableRef})
    @ SpineInterface C:\Users\Manuel_Marin\Codes\SpineInterface.jl\src\update_model.jl:405
  [7] +
    @ C:\Users\Manuel_Marin\Codes\SpineInterface.jl\src\update_model.jl:409 [inlined]
  [8] -
    @ C:\Users\Manuel_Marin\Codes\SpineInterface.jl\src\update_model.jl:412 [inlined]
  [9] sub_mul
    @ C:\Users\Manuel_Marin\.julia\packages\MutableArithmetics\NIXlP\src\MutableArithmetics.jl:38 [inlined]
 [10] operate!
    @ C:\Users\Manuel_Marin\Codes\SpineInterface.jl\src\update_model.jl:450 [inlined]
 [11] operate_fallback!!
    @ C:\Users\Manuel_Marin\.julia\packages\MutableArithmetics\NIXlP\src\interface.jl:601 [inlined]
 [12] operate!!(op::typeof(MutableArithmetics.sub_mul), x::JuMP.GenericAffExpr{SpineInterface.Call, JuMP.VariableRef}, args::SpineInterface.Call)
    @ MutableArithmetics C:\Users\Manuel_Marin\.julia\packages\MutableArithmetics\NIXlP\src\rewrite.jl:93
 [13] macro expansion
    @ C:\Users\Manuel_Marin\.julia\packages\MutableArithmetics\NIXlP\src\rewrite.jl:321 [inlined]
 [14] macro expansion
    @ C:\Users\Manuel_Marin\.julia\packages\JuMP\R53zo\src\macros.jl:721 [inlined]
 [15] #1284
    @ C:\Users\Manuel_Marin\Codes\SpineOpt.jl\src\constraints\constraint_units_available.jl:38 [inlined]
 [16] (::SpineOpt.var"#1284#1285"{JuMP.Model, SpineInterface._StartRef, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, SpineInterface.Call}, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}})(::NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}})
    @ SpineOpt .\none:0
 [17] iterate
    @ .\generator.jl:47 [inlined]
 [18] _all(f::Base.var"#372#374", itr::Base.Generator{Vector{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}}, SpineOpt.var"#1284#1285"{JuMP.Model, SpineInterface._StartRef, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, SpineInterface.Call}, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}}}, #unused#::Colon)
    @ Base .\reduce.jl:1282
 [19] all
    @ .\reduce.jl:1278 [inlined]
 [20] Dict(kv::Base.Generator{Vector{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}}, SpineOpt.var"#1284#1285"{JuMP.Model, SpineInterface._StartRef, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, SpineInterface.Call}, Dict{NamedTuple{(:unit, :stochastic_scenario, :t), Tuple{SpineInterface.Object, SpineInterface.Object, SpineInterface.TimeSlice}}, JuMP.VariableRef}}})
    @ Base .\dict.jl:111
 [21] add_constraint_units_available!(m::JuMP.Model)
    @ SpineOpt C:\Users\Manuel_Marin\Codes\SpineOpt.jl\src\constraints\constraint_units_available.jl:37
 [22] macro expansion
    @ .\timing.jl:273 [inlined]
 [23] macro expansion
    @ C:\Users\Manuel_Marin\Codes\SpineOpt.jl\src\util\misc.jl:51 [inlined]
 [24] _add_constraints!(m::JuMP.Model; add_constraints::SpineOpt.var"#68#76", log_level::Int64)
    @ SpineOpt C:\Users\Manuel_Marin\Codes\SpineOpt.jl\src\run_spineopt_standard.jl:215
 [25] macro expansion
    @ .\timing.jl:273 [inlined]
 [26] macro expansion
    @ C:\Users\Manuel_Marin\Codes\SpineOpt.jl\src\util\misc.jl:51 [inlined]
 [27] init_model!(m::JuMP.Model; add_user_variables::Function, add_constraints::Function, log_level::Int64)
    @ SpineOpt C:\Users\Manuel_Marin\Codes\SpineOpt.jl\src\run_spineopt_standard.jl:79
 [28] init_model!
    @ C:\Users\Manuel_Marin\Codes\SpineOpt.jl\src\run_spineopt_standard.jl:75 [inlined]
 [29] rerun_spineopt_benders!(m::JuMP.Model, url_out::String; add_user_variables::Function, add_constraints::Function, log_level::Int64, optimize::Bool, update_names::Bool, alternative::String, write_as_roll::Int64, resume_file_path::Nothing, run_kernel::typeof(SpineOptReform._run_lt_storage_kernel!))
    @ SpineOpt C:\Users\Manuel_Marin\Codes\SpineOpt.jl\src\run_spineopt_benders.jl:43
 [30] invokelatest(::Any, ::Any, ::Vararg{Any}; kwargs::Base.Pairs{Symbol, Any, NTuple{9, Symbol}, NamedTuple{(:add_user_variables, :add_constraints, :log_level, :optimize, :update_names, :alternative, :write_as_roll, :resume_file_path, :run_kernel), Tuple{SpineOpt.var"#67#75", SpineOpt.var"#68#76", Int64, Bool, Bool, String, Int64, Nothing, typeof(SpineOptReform._run_lt_storage_kernel!)}}})
    @ Base .\essentials.jl:821
 [31] rerun_spineopt(url_out::String; mip_solver::Nothing, lp_solver::Nothing, add_user_variables::Function, add_constraints::Function, log_level::Int64, optimize::Bool, update_names::Bool, alternative::String, write_as_roll::Int64, resume_file_path::Nothing, use_direct_model::Bool, run_kernel::typeof(SpineOptReform._run_lt_storage_kernel!), handle_window_about_to_solve::SpineOpt.var"#69#77", handle_window_solved::SpineOpt.var"#70#78")
    @ SpineOpt C:\Users\Manuel_Marin\Codes\SpineOpt.jl\src\run_spineopt.jl:278
 [32] rerun_spineopt
    @ C:\Users\Manuel_Marin\Codes\SpineOpt.jl\src\run_spineopt.jl:251 [inlined]
 [33] run_spineopt_lt_storage(url_in::String, url_out::String; upgrade::Bool, log_level::Int64, filters::Dict{String, String}, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ SpineOptReform C:\Users\Manuel_Marin\Codes\SpineOptReform\src\SpineOptReform.jl:31
 [34] top-level scope
    @ C:\Users\Manuel_Marin\Codes\just_run_spineopt_lt_storage.jl:11
 [35] include(fname::String)
    @ Base.MainInclude .\client.jl:478
 [36] top-level scope
    @ REPL[4]:1
in expression starting at C:\Users\Manuel_Marin\Codes\just_run_spineopt_lt_storage.jl:11
```

**Comment by DillonJ** [comment]
@manuelma Did you load the new template?

I think this is because units_unavailable[args... , _default=0] returns nothing. 

If you load the template it should work

**Comment by manuelma** [comment]
> @manuelma Did you load the new template?
> 
> I think this is because `units_unavailable[(args... , _default=0)]` returns nothing instead of zero.
> 
> If you load the template it should work

Thank you @DillonJ that makes sense. I can see now how this is frustrating for the user who wants to run their model. They will see a long traceback and probably won't associate that with upgrade=false or loading the template. We need a better solution and we should have in place before we merge any more breaking changes to the template I think.

File: src/constraints/constraint_units_out_of_service_contiguity.jl [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        for u in indices(scheduled_outage_duration)
+        for (u, t) in unit_time_indices(m; unit=u)
+        for path in active_stochastic_paths(m, past_units_out_of_service_indices(m, u, anything, t, scheduled_outage_duration))
+    )
+end
+
+function past_units_out_of_service_indices(m, u, s, t, min_time)
+    t0 = _analysis_time(m)
+    units_on_indices(
+        m;
```
Comment (manuelma) [inline src/constraints/constraint_units_out_of_service_contiguity.jl:71]: Shouldn't this be `units_out_of_service_indices` instead @DillonJ

Comment (DillonJ) [reply src/constraints/constraint_units_out_of_service_contiguity.jl:71]: they're the same really, but I guess, strictly speaking, it should

**Comment by manuelma** [comment]
Another potential issue here is that one needs JuMP v1.20.0 for things to work. I had v1.8.2 and was getting errors with the new constraints. I guess we'd need to be more careful in the future and always add the JuMP version to the compat section of Project.toml? What do you think @DillonJ ?

File: src/constraints/constraint_units_out_of_service_contiguity.jl (hunk 10 above)
Comment (manuelma) [reply src/constraints/constraint_units_out_of_service_contiguity.jl:71]: Thanks, will change it in master directly I think.

**Pull request merged**